test/libsolidity/NatspecJSONTest.cpp

```diff
@@ -0,0 +1,213 @@
+		if (strippedLine.empty())
+			continue;
+
+		auto [contractName, kind] = parseExpectationHeader(strippedLine);
+
+		string rawJSON = extractExpectationJSON(_stream);
+		string jsonErrors;
+		Json::Value parsedJSON;
+		bool jsonParsingSuccessful = jsonParseStrict(rawJSON, parsedJSON, &jsonErrors);
+		if (!jsonParsingSuccessful)
+			BOOST_THROW_EXCEPTION(runtime_error(fmt::format(
+				"Malformed JSON in {} expectation for contract {}.\n"
+				"Note that JSON expectations must be pretty-printed to be split correctly. "
+				"The object is assumed to and at the first unindented closing brace.\n"
+				"{}",
+				toString(kind),
+				contractName,
+				rawJSON
+			)));
+
+		m_expectedNatspecJSON[string(contractName)][kind] = parsedJSON;
+	}
+}
+
+bool NatspecJSONTest::expectationsMatch()
+{
+	// NOTE: Comparing pretty printed Json::Values to avoid using its operator==, which fails to
+	// compare equal numbers as equal. For example, for 'version' field the value is sometimes int,
+	// sometimes uint and they compare as different even when both are 1.
+	return
+		SyntaxTest::expectationsMatch() &&
+		prettyPrinted(obtainedNatspec()) == prettyPrinted(m_expectedNatspecJSON);
+}
+
+void NatspecJSONTest::printExpectedResult(ostream& _stream, string const& _linePrefix, bool _formatted) const
+{
+	SyntaxTest::printExpectedResult(_stream, _linePrefix, _formatted);
+	if (!m_expectedNatspecJSON.empty())
+	{
+		_stream << _linePrefix << "----" << endl;
+		printIndented(_stream, formatNatspecExpectations(m_expectedNatspecJSON), _linePrefix);
+	}
+}
+
+void NatspecJSONTest::printObtainedResult(ostream& _stream, string const& _linePrefix, bool _formatted) const
+{
+	SyntaxTest::printObtainedResult(_stream, _linePrefix, _formatted);
+
+	NatspecMap natspecJSON = obtainedNatspec();
+	if (!natspecJSON.empty())
+	{
+		_stream << _linePrefix << "----" << endl;
+		// TODO: Diff both versions and highlight differences.
+		// We should have a helper for doing that in newly defined test cases without much effort.
+		printIndented(_stream, formatNatspecExpectations(natspecJSON), _linePrefix);
+	}
+}
+
+tuple<string_view, NatspecJSONKind> NatspecJSONTest::parseExpectationHeader(string_view _line)
+{
+	for (NatspecJSONKind kind: {NatspecJSONKind::Devdoc, NatspecJSONKind::Userdoc})
+	{
+		string kindSuffix = " " + toString(kind);
+		if (boost::algorithm::ends_with(_line, kindSuffix))
+			return {_line.substr(0, _line.size() - kindSuffix.size()), kind};
+	}
+
+	BOOST_THROW_EXCEPTION(runtime_error(
+		"Natspec kind (devdoc/userdoc) not present in the expectation: "s.append(_line)
+	));
+}
+
+string NatspecJSONTest::extractExpectationJSON(istream& _stream)
+{
+	string rawJSON;
+	string line;
+	while (getline(_stream, line))
+	{
+		string_view strippedLine = expectLinePrefix(line);
+		rawJSON += strippedLine;
+		rawJSON += "\n";
+
+		if (strippedLine.empty())
+			continue;
+
+		if (boost::algorithm::starts_with(strippedLine, "}"))
```

Comment:
So, since it's you, I'm sure this is right, but I'm not quite sure I understand what this is doing? I assume we're consuming lines until we encounter a closing brace, which should indicate that we've read the whole JSON object, but wouldn't this just break at the first encountered closing brace? E.g.

```
// {
//     "kind": "dev",
//     "methods":
//     {
//         "g(uint256)":
//         {
//             "custom:since": "2014"
//         }  <--- break here
//     },
//     "version": 1
// }
```

Reply:
No, because we're not removing all whitespace. Only the initial `// ` so the content we see is still properly indented.

If it's pretty printed, as it should, it will work fine. If it's not, JSON parsing will fail and you'll run into an assert that tells you to pretty print it.